tests/utils/log_reader.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2014-2020 Scalyr Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import unicode_literals
+from __future__ import print_function
+from __future__ import absolute_import
+
+if False:
+    from typing import Optional
+
+import re
+import threading
+import time
+from io import open
+import collections
+
+import six
+
+
+class LogReader(threading.Thread):
+    """Reader that allows to read file in separate thread and read and wait for new lines"""
+
+    def __init__(self, file_path):
+        super(LogReader, self).__init__()
+        self.daemon = True
+        self._file = open(six.text_type(file_path), "r")
+        self._stop_event = threading.Event()
+        self._lines_lock = threading.Lock()
+        self._start_cv = threading.Condition()
+        self._has_data = False
+        self._lines = collections.deque(maxlen=1000)
+        self._wait_for_line_cv = threading.Condition()
+
+        self._has_new_line = False
+
+    def handle_new_line(self, line):
+        with self._lines_lock:
+            self._lines.append(line)
+        with self._wait_for_line_cv:
+            self._has_new_line = True
+            self._wait_for_line_cv.notify()
+
+    def run(self):
+        def get_line():
+            while not self._stop_event.is_set():
+                _line = self._file.readline()
+                if _line:
+                    return _line
+                time.sleep(0.1)
+
+        get_line()
+        with self._start_cv:
+            self._has_data = True
+            self._start_cv.notify()
+        self._file.seek(0)
+        while not self._stop_event.is_set():
+            line = get_line()
+            self.handle_new_line(line)
+
+    def start(self, wait_for_data=True):
+        super(LogReader, self).start()
+        if not wait_for_data:
+            return
+        with self._start_cv:
+            while not self._has_data:
+                self._start_cv.wait(0.1)
+        return
+
+    @property
+    def lines(self):
+        with self._lines_lock:
+            return list(self._lines)
+
+    def wait_for_new_line(self):
```

Comment:
We should probably add support for timeout here, right?

In case there is a misconfiguration or test not working correctly or similar so we prevent tests from waiting / hanging util Circle CI job timeout (10 minutes) is reached.